perf_bench_test.go

```diff
@@ -24,127 +24,619 @@ package main
+		cpuList[1] = 15
+	}
+
+	dumpMetrics(t, metrics, metrFile)
+
+	profile.PlotCVS(1, *benchDir, metrFile, "the number of cores")
+
+	tearDownVMs(t, images, 2, isSyncOffload)
+}
 
-		// For future use
-		// createSnapshots(t, concurrency, vmID, imageName, isSyncOffload)
-		// log.Info("Snapshot created")
-		// createRecords(t, concurrency, vmID, imageName, isSyncOffload)
-		// log.Info("Record done")
+func TestBindSocket(t *testing.T) {
+	var (
+		servedTh      uint64
+		pinnedFuncNum int
+		isSyncOffload bool = true
+		testImage          = []string{"vhiveease/helloworld:var_workload"}
+	)
 
-		funcs = append(funcs, funcName)
+	type testCase struct {
+		vmNum    int
+		expected []string
 	}
 
-	// Boot VMs
-	for i := 0; i < *vmNum; i++ {
+	cases := []testCase{
+		{vmNum: 1, expected: []string{"13"}},
+		{vmNum: 4, expected: []string{"13", "1,3,5,7,9,11,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43," +
+			"47,49,51,53,55,57,59,61,63"}},
+	}
+
+	funcPool = NewFuncPool(!isSaveMemoryConst, servedTh, pinnedFuncNum, isTestModeConst)
+	*isBindSocket = true
+
+	for _, tCase := range cases {
+		testName := fmt.Sprintf("vmNum=%d", tCase.vmNum)
+		t.Run(testName, func(t *testing.T) {
+			bootVMs(t, testImage, 0, tCase.vmNum)
+
+			pidBytes, err := getFirecrackerPid()
+			require.NoError(t, err, "Cannot get Firecracker PID")
+			vmPidList := strings.Split(string(pidBytes), " ")
+
+			cpuBytes, err := exec.Command("taskset", "-cp", strings.TrimSpace(vmPidList[0])).Output()
+			require.NoError(t, err, "Cannot get CPU affinity")
+			cpuAffinity := strings.TrimSpace(strings.Split(string(cpuBytes), ":")[1])
+			require.Equal(t, tCase.expected, cpuAffinity[0], "VM was not binded to CPU 13")
+			for _, vm := range vmPidList[1:] {
+				vm = strings.TrimSpace(vm)
+				cpuBytes, err := exec.Command("taskset", "-cp", vm).Output()
+				require.NoError(t, err, "Cannot get CPU affinity")
+				cpuAffinity := strings.TrimSpace(strings.Split(string(cpuBytes), ":")[1])
+				require.Equal(t, tCase.expected, cpuAffinity[1], "VM was not binded to socket 1")
+			}
+
+			tearDownVMs(t, testImage, tCase.vmNum, isSyncOffload)
+		})
+	}
+}
+
+// bootVMs boots a range of VMs with given images
+func bootVMs(t *testing.T, images []string, startVMID, endVMID int) {
+	for i := startVMID; i < endVMID; i++ {
 		vmIDString := strconv.Itoa(i)
-		_, err := funcPool.AddInstance(vmIDString, images[funcs[i%len(funcs)]])
+		_, err := funcPool.AddInstance(vmIDString, images[i%len(images)])
 		require.NoError(t, err, "Function returned error")
 	}
 
-	if !*isWithCache && *isColdStart {
-		log.Info("Profile cold start")
-		dropPageCache()
+	if *isBindSocket {
+		log.Debugf("Binding socket 1")
+		err := bindSocket()
+		require.NoError(t, err, "Bind Socket returned error")
 	}
+}
 
-	ticker := time.NewTicker(timeInterval)
-	var vmGroup sync.WaitGroup
-	done := make(chan bool, 1)
+// loadAndProfile loads from 5% to 100% of input target RPS and profile counters iteratively
```

Comment:
the step is configurable!

Reply:
loads -> issues requests at B% to M% of the maximum RPS (measured separately), where B is ... and M is ...
profile counters -> collects counters (or collects the profile)

Reply:
Done